Use Map instead of Array to register event listeners

rad-pat opened the pull request:
To fix https://github.com/qooxdoo/qooxdoo/issues/10620, change the storage of event listener callbacks from an Array to using a Map, keyed by the unique Id. I think I've made sure the exposed APIs return the same as before.

The real speed-up comes from them storing the unique IDs for the added listeners and then using those for the removal from the Map. This makes a huge difference where there are many listeners in the Array (Map now) for the `qx.theme.manager.Meta` or the `qx.locale.Manager` when dynamic theming or locales are enabled.

rad-pat committed:
Use Map over Array for storing listeners in Event Manager

cboulanger reviewed (approved):
Seems sensible - I trust that this has been extensively tested not to break anything.

rad-pat commented:
> Seems sensible - I trust that this has been extensively tested not to break anything.

The tests seemed to do a reasonable job of covering the changes, but we can run with this for a while before merging to ensure that nothing is broken.

goldim reviewed (changes requested):
I think after removing listener you have to assign null/undefined to private member.

johnspackman commented:
Hi,

Thank you for your email - I will be on annual leave and unavailable on Tuesday 12th and Wednesday 13th November 2024.
  
 If you require assistance, please contact your IT department in the first instance, who will be able to pass on any questions to the appropriate person to handle your query in my absence.
  
 I will get back to you as soon as possible, but if you have an emergency support issue please describe your issue in detail in email and then send a text message to the mobile number you have for me.

Regards
John Spackman

rad-pat commented:
> I think after removing listener you have to assign null/undefined to private member.
When looking at assigning null to the stored listenerIds, it seems that other listener ids are not cleared. Presumably they are garbage collected just fine as they're just a string.

goldim commented:
> > I think after removing listener you have to assign null/undefined to private member.
> 
> When looking at assigning null to the stored listenerIds, it seems that other listener ids are not cleared. Presumably they are garbage collected just fine as they're just a string.

You want to say that they won't be reusable?

goldim commented:
Just looked through, yes, most cases - it is cleared in destructor but for example see https://github.com/qooxdoo/qooxdoo/blob/2cab79dc2fa95ab1c6e289c5a01a8a2dcf52c732/source/class/qx/ui/menu/AbstractButton.js . In this file u can twice call `_on changeCommand` and in the second time it also will have string value and `removeListener` will be called.

goldim commented:
Actually there is a strange logic in this method. There is variable checking on null/undefined and if yes it returns control but if you see the next code of this method there are still checking on null/undefined. I think this is for new issue. I think removing listener should be before the first return otherwise there is no point to remove it.

goldim commented:
Also you changed API of method `getAllListeners` - returns JS object but now it is Map. It is broken BC. And you have to keep tests untouched. Try to save return value, transform Map into JS object.

goldim commented:
@rad-pat I've just read the issue conversation and don't know what to say. I didn't participate in it and don't see conclusion. Is there way to save somehow return value of the method?

oetiker reviewed (commented):
thanks @goldim for investigating

rad-pat commented:
> Also you changed API of method `getAllListeners` - returns JS object but now it is Map. It is broken BC. And you have to keep tests untouched. Try to transform Map into JS object when returning value.

The `getAllListeners` was marked as internal, presumably created for the purpose of checking what listeners exist before and after the tests are ran. I've updated all the test instances where it is used to cope with the Map instead, do you want me to revert those changes to tests? I thought because it was internal it would be ok to update that.

oetiker commented:
> > Also you changed API of method `getAllListeners` - returns JS object but now it is Map. It is broken BC. And you have to keep tests untouched. Try to transform Map into JS object when returning value.
> 
> The `getAllListeners` was marked as internal, presumably created for the purpose of checking what listeners exist before and after the tests are ran. I've updated all the test instances where it is used to cope with the Map instead, do you want me to revert those changes to tests? I thought because it was internal it would be ok to update that.

Unfortunately the `@internal` designation does not mean there is no stable API. The only things that ARE internal are the `__variables` and `__methods`. Unfortunately with this method we are exposing the internal data structure ... 

If you look through the original bug discussion you can see some sample code I created to show how the data interface can be kept ...

rad-pat commented:
OK, I'll hopefully get back on this at some point next week and see if I can fix it up. 👍

rad-pat committed:
Remove Dynamic Theme & Locale listeners by Id

rad-pat commented:
@oetiker, @goldim I have revised the changes, removed changes to the tests. Added Proxy objects to the results of `getAllListeners` and `getListeners` methods to help monitor any changes made to those arrays.

goldim commented:
@rad-pat I will check it out on these days or on this weekend.

hkollmann commented:
@oetiker: Is this ok for you now?
